File: hash_table.py
class HashNode:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = None

class HashTable:
    def __init__(self, capacity=10):
        self.capacity = capacity
        self.buckets = [None] * capacity
        self.size = 0
    
    def _hash(self, key):
        return hash(key) % self.capacity
    
    def put(self, key, value):
        index = self._hash(key)
        head = self.buckets[index]
        
        # Check if key already exists
        current = head
        while current:
            if current.key == key:
                current.value = value
                return
            current = current.next
        
        # Add new node at the beginning
        new_node = HashNode(key, value)
        new_node.next = head
        self.buckets[index] = new_node
        self.size += 1
    
    def get(self, key):
        index = self._hash(key)
        head = self.buckets[index]
        
        while head:
            if head.key == key:
                return head.value
            head = head.next
        
        return None
    
    def remove(self, key):
        index = self._hash(key)
        head = self.buckets[index]
        prev = None
        
        while head:
            if head.key == key:
                if prev:
                    prev.next = head.next
                else:
                    self.buckets[index] = head.next
                self.size -= 1
                return True
            prev = head
            head = head.next
        
        return False
    
    def __len__(self):
        return self.size
    
    def is_empty(self):
        return self.size == 0

if __name__ == "__main__":
    hash_table = HashTable()
    
    hash_table.put("apple", 5)
    hash_table.put("banana", 3)
    hash_table.put("orange", 8)
    
    print(f"apple: {hash_table.get('apple')}")
    print(f"banana: {hash_table.get('banana')}")
    print(f"grape: {hash_table.get('grape')}")
    
    hash_table.remove("banana")
    print(f"After removing banana: {hash_table.get('banana')}")
    print(f"Size: {len(hash_table)}")

File: Java/data_structures/HashTable.java
import java.util.*;

class HashNode {
    String key;
    int value;
    HashNode next;
    
    HashNode(String key, int value) {
        this.key = key;
        this.value = value;
    }
}

public class HashTable {
    private HashNode[] buckets;
    private int capacity;
    private int size;
    
    public HashTable(int capacity) {
        this.capacity = capacity;
        this.buckets = new HashNode[capacity];
        this.size = 0;
    }
    
    private int hash(String key) {
        return Math.abs(key.hashCode()) % capacity;
    }
    
    public void put(String key, int value) {
        int index = hash(key);
        HashNode head = buckets[index];
        
        // Check if key already exists
        while (head != null) {
            if (head.key.equals(key)) {
                head.value = value;
                return;
            }
            head = head.next;
        }
        
        // Add new node at the beginning
        HashNode newNode = new HashNode(key, value);
        newNode.next = buckets[index];
        buckets[index] = newNode;
        size++;
    }
    
    public Integer get(String key) {
        int index = hash(key);
        HashNode head = buckets[index];
        
        while (head != null) {
            if (head.key.equals(key)) {
                return head.value;
            }
            head = head.next;
        }
        
        return null;
    }
    
    public boolean remove(String key) {
        int index = hash(key);
        HashNode head = buckets[index];
        HashNode prev = null;
        
        while (head != null) {
            if (head.key.equals(key)) {
                if (prev != null) {
                    prev.next = head.next;
                } else {
                    buckets[index] = head.next;
                }
                size--;
                return true;
            }
            prev = head;
            head = head.next;
        }
        
        return false;
    }
    
    public int size() {
        return size;
    }
    
    public boolean isEmpty() {
        return size == 0;
    }
    
    public static void main(String[] args) {
        HashTable hashTable = new HashTable(10);
        
        hashTable.put("apple", 5);
        hashTable.put("banana", 3);
        hashTable.put("orange", 8);
        
        System.out.println("apple: " + hashTable.get("apple"));
        System.out.println("banana: " + hashTable.get("banana"));
        System.out.println("grape: " + hashTable.get("grape"));
        
        hashTable.remove("banana");
        System.out.println("After removing banana: " + hashTable.get("banana"));
        System.out.println("Size: " + hashTable.size());
    }
}